algebird-test/src/main/scala/com/twitter/algebird/macros/ArbitraryCaseClassMacro.scala

```diff
@@ -0,0 +1,53 @@
+package com.twitter.algebird.macros
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+import scala.reflect.runtime.universe._
+
+import com.twitter.algebird._
+
+import org.scalacheck.{ Gen, Arbitrary }
+
+object ArbitraryCaseClassMacro {
+  def caseClassGen[T](c: Context)(implicit T: c.WeakTypeTag[T]): c.Expr[Gen[T]] = {
+    import c.universe._
+
+    val tpe = weakTypeOf[T]
+
+    val isCaseClass = tpe.typeSymbol.isClass && tpe.typeSymbol.asClass.isCaseClass
+    if (!isCaseClass)
+      c.abort(c.enclosingPosition, s"$T is not a clase class")
+
+    val params = tpe.declarations.collect {
+      case m: MethodSymbol if m.isCaseAccessor => m
+    }.toList
+
+    val getsList = params.zipWithIndex.map {
+      case (param, i) =>
+        fq"${param.name} <- _root_.org.scalacheck.Arbitrary.arbitrary[${param.returnType}]"
+    }
+
+    val paramsList = params.map(param => q"${param.name.asInstanceOf[TermName]}")
+    val companion = tpe.typeSymbol.companionSymbol
+
+    val res = q"""
+    for ( ..$getsList ) yield $companion.apply(..$paramsList)
+    """
+
+    c.Expr[Gen[T]](res)
+  }
+
+  def caseClassArbitrary[T](c: Context)(implicit T: c.WeakTypeTag[T]): c.Expr[Arbitrary[T]] = {
```

Comment:
this is hugely useful. Nice work. It should almost go in its own package, but we can leave it here for now.

This will be really useful at Twitter to have this (and a similar one for scrooge (and java thrift) if possible).